axiom/nr_txn.c

```diff
@@ -3293,6 +3297,90 @@ bool nr_txn_log_decorating_enabled(nrtxn_t* txn) {
   return false;
 }
 
+#define ENSURE_LOG_LEVEL_NAME(level_name) \
+  (NULL == level_name ? "UNKNOWN" : level_name)
+
+static void log_event_set_linking_metadata(nr_log_event_t* e,
+                                           nrtxn_t* txn,
+                                           nrapp_t* app) {
+  char* trace_id = NULL;
+  char* span_id = NULL;
+
+  if (NULL == e) {
+    return;
+  }
+
+  if (nrlikely(txn)) {
+    trace_id = nr_txn_get_current_trace_id(txn);
+    nr_log_event_set_trace_id(e, trace_id);
+    nr_free(trace_id);
+
+    span_id = nr_txn_get_current_span_id(txn);
+    nr_log_event_set_span_id(e, span_id);
+    nr_free(span_id);
+
+    nr_log_event_set_entity_name(e, txn->primary_app_name);
+  }
+
+  if (nrlikely(app)) {
+    nr_log_event_set_hostname(e, nr_app_get_host_name(app));
+    nr_log_event_set_guid(e, nr_app_get_entity_guid(app));
+  }
+}
+
+static nr_log_event_t* log_event_create(const char* log_level_name,
+                                        const char* log_message,
+                                        nrtime_t timestamp,
+                                        nrtxn_t* txn,
+                                        nrapp_t* app) {
+  nr_log_event_t* e = nr_log_event_create();
+  nr_log_event_set_log_level(e, ENSURE_LOG_LEVEL_NAME(log_level_name));
+  nr_log_event_set_message(e, log_message);
+  nr_log_event_set_timestamp(e, timestamp);
+
+  log_event_set_linking_metadata(e, txn, app);
+
+  return e;
+}
+
+static void nr_txn_add_log_event(nrtxn_t* txn,
+                                 const char* log_level_name,
+                                 const char* log_message,
+                                 nrtime_t timestamp,
+                                 nrapp_t* app) {
+  nr_random_t* rnd = NULL;
+  nr_log_event_t* e = NULL;
+  bool event_dropped = false;
+
+  if (!nr_txn_log_forwarding_enabled(txn)) {
+    return;
+  }
+
+  if (nr_strempty(log_message)) {
+    nrl_debug(NRL_TXN, "%s: not forwarding log event with empty log message",
+              __func__);
+    return;
+  }
+
+  rnd = nr_random_create();
+  nr_random_seed(rnd, timestamp);
+
+  e = log_event_create(log_level_name, log_message, timestamp, txn, app);
+  if (NULL == e) {
+    nrl_debug(NRL_TXN, "%s: failed to create log event", __func__);
+    event_dropped = true;
+  } else {
+    event_dropped = nr_log_events_add_event(txn->log_events, e, rnd);
+  }
+  nr_log_event_destroy(&e);
+
+  nr_random_destroy(&rnd);
+
+  if (event_dropped) {
```

Comment:
Does this metrics creation depend on if log metrics are enabled?

Reply:
It does not. The spec clearly states it is part of log forwarding feature.

Reply:
I'll ask for clarification on this behavior - I agree as written it does seem like this would currently be created even if the metrics flag is set to disabled (which seems odd/wrong).

Reply:
This is correct as implemented to resolving this conversation.